src/main/java/com/google/sps/servlets/SurveyInputServlet.java

```diff
@@ -0,0 +1,122 @@
+    private static final String USER_DEED_DESCRIPTION = "description";
+    private static final String USER_DEED_VOTES = "votes";
+    private static final String USER_DEED_TIMESTAMP = "timestamp";
+
+    private static final String WEBAPP_SURVEY_INPUT = "/surveyInput.html";
+    private static final String JSON_CONTENT_TYPE = "application/json;";
+    private static final long ZERO = 0;
+
+    private static DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    private Gson gson = new Gson();
+ 
+ 
+    /*
+     * Grabs any previous ideas from the datastore and posts them as a Json
+     */
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException{
+        Query query = new Query(USER_DEED_ENTITY).addSort(USER_DEED_VOTES, SortDirection.DESCENDING);
+      
+        PreparedQuery userIdeas = datastore.prepare(query);
+
+        List<SurveyInput> ideasList = new ArrayList<>();
+        for (Entity entity : userIdeas.asIterable()) {
+            
+            long id = entity.getKey().getId();
+            String newIdea = (String) entity.getProperty(USER_DEED_TITLE);
+            String newIdeaDescription = (String) entity.getProperty(USER_DEED_DESCRIPTION);
+            long newIdeaVotes = (long) entity.getProperty(USER_DEED_VOTES);
+            long timestamp = (long) entity.getProperty(USER_DEED_TIMESTAMP);
+
+            SurveyInput ideaObject = new SurveyInput(id, newIdea, newIdeaDescription, newIdeaVotes, timestamp);
+            ideasList.add(ideaObject);
+        }
+
+        response.setContentType(JSON_CONTENT_TYPE);
+        response.getWriter().println(gson.toJson(ideasList));
+    }
+ 
+    /*
+     * Allows a user to enter a new idea, makes a new entity for that idea, and puts it in the 
+     * datastore
+     */
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String newIdeaTitle = request.getParameter(FORM_IDEA_TITLE);
+        String newIdeaDescription = request.getParameter(FORM_IDEA_DESCRIPTION);
+        long timestamp = System.currentTimeMillis(); 
+
+        if (isValid(newIdeaTitle)) {
+            if (isValid(newIdeaDescription)) {
+                Entity ideaEntity = new Entity(USER_DEED_ENTITY);
+                ideaEntity.setProperty(USER_DEED_TITLE, newIdeaTitle);
+                ideaEntity.setProperty(USER_DEED_DESCRIPTION, newIdeaDescription);
+                ideaEntity.setProperty(USER_DEED_VOTES, ZERO);
+                ideaEntity.setProperty(USER_DEED_TIMESTAMP, timestamp);
+
+                datastore.put(ideaEntity);
+            }
+      }
+        response.sendRedirect(WEBAPP_SURVEY_INPUT); // Sends users back to the survey input page
+    }
+
+    /**
+     * Checks if the user's string is not empty
+     */
+    public boolean isValid(String comment) {
+      return !comment.isEmpty() && !comment.isBlank() && !(comment == null);
```

Comment:
If `comment` is `null`, you will have a runtime crash here. Always have the nullability check first, this way you can avert a `NullPointerException`.  Also, recommend using `comment != null` over `!(comment == null)`. The latter is confusing to read

Reply:
True! Let me change that!